bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/AbstractProcessor.java

```diff
@@ -0,0 +1,156 @@
+public abstract class AbstractProcessor {
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /*
+     * Used to extract remote sensor data from the data event Json
+     */
+    protected RemoteSensor remoteSensor = new RemoteSensor();
+
+    /*
+     * Updates the info channels (i.e. name and location) for a station
+     */
+    public abstract void processInfoUpdate(AmbientWeatherStationHandler handler, String station, String name,
+            String location);
+
+    /*
+     * Updates the weather data channels for a station
+     */
+    public abstract void processWeatherData(AmbientWeatherStationHandler handler, String station, String jsonData);
+
+    // @formatter:off
+    private static final String[] UV_INDEX = {
+        "LOW",
+        "LOW",
+        "LOW",
+        "MODERATE",
+        "MODERATE",
+        "MODERATE",
+        "HIGH",
+        "HIGH",
+        "VERY HIGH",
+        "VERY HIGH",
+        "VERY HIGH",
+        "EXTREME",
+        "EXTREME",
+        "EXTREME",
+        "EXTREME",
+        "EXTREME"
+    };
+    // @formatter:on
+
+    /*
+     * Helper function called by processor to convert UTC time milliseconds to local time
+     */
+    public DateTimeType getLocalDateTimeType(long dateTimeMillis, ZoneId zoneId) {
+        Instant instant = Instant.ofEpochMilli(dateTimeMillis);
+        ZonedDateTime localDateTime = instant.atZone(zoneId);
+        DateTimeType dateTimeType = new DateTimeType(localDateTime);
+        return dateTimeType;
+    }
+
+    /*
+     * Helper function called by processor to convert UTC time string to local time
+     * Input string is of form 2018-12-02T10:47:00.000Z
+     */
+    public DateTimeType getLocalDateTimeType(String dateTimeString, ZoneId zoneId) {
+        DateTimeType dateTimeType;
+        try {
+            Instant instant = Instant.parse(dateTimeString);
+            ZonedDateTime localDateTime = instant.atZone(zoneId);
+            dateTimeType = new DateTimeType(localDateTime);
+        } catch (DateTimeParseException e) {
+            logger.debug("Error parsing date/time string: {}", e.getMessage());
+            dateTimeType = new DateTimeType();
+        } catch (IllegalArgumentException e) {
+            logger.debug("Error converting to DateTimeType: {}", e.getMessage());
+            dateTimeType = new DateTimeType();
+        }
+        return dateTimeType;
+    }
+
+    /*
+     * Convert the UV Index integer value to a string representation
+     */
+    public String convertUVIndexToString(int uvIndex) {
+        if (uvIndex < 0 || uvIndex >= UV_INDEX.length) {
+            return "UNKNOWN";
+        }
+        return UV_INDEX[uvIndex];
+    }
+
+    /*
+     * Convert the wind directions in degrees to a string representation
+     */
+    public String convertWindDirectionToString(double windDirectionDegrees) {
+        String result = "UNKNOWN";
+        if (windDirectionDegrees >= 0 && windDirectionDegrees <= 11.25) {
```

Comment:
Can you make add those entries into a `Treemap` please. See https://docs.oracle.com/javase/6/docs/api/java/util/TreeMap.html#ceilingKey(K). 

A data structure will not only be faster but also more easy to extend or adapt in the future (that is of course not necessary in this case. It's still faster)

Reply:
@davidgraeff 

I found an approach used in another binding ([synopanalyzer](https://github.com/openhab/openhab2-addons/blob/1fd57c40758fde6f8eca936f285ef0bf06894c79/bundles/org.openhab.binding.synopanalyzer/src/main/java/org/openhab/binding/synopanalyzer/internal/handler/SynopAnalyzerHandler.java#L179)). While it doesn't use a TreeMap, it does a simple index of an array (albeit at the cost of some upfront math). It does need some minor modification to avoid a potential AIOOB exception. WDYT?
```
private String getWindDirection(int degree) {
    double step = 360.0 / WIND_DIRECTIONS.length;
    double b = Math.floor((degree + (step / 2.0)) / step);
    return WIND_DIRECTIONS[(int) (b % WIND_DIRECTIONS.length)];
}

public static final String[] WIND_DIRECTIONS = new String[] { "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE", "S",
        "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW" };
```

Reply:
That's fine as well, the data is continuous so an array with computing the index should work. The modulo makes sure that we do not get out of bounds exceptions, doesn't it?

Reply:
> That's fine as well,

Ok, thanks.

> The modulo makes sure that we do not get out of bounds exceptions, doesn't it?

It sure does!

Reply:
BTW, to use this piece of code, do I need to add an Also-by to the PR and/or commit?

Reply:
Nope. Not because of 5 lines of code without any special algorithm inside